# Đồ án cuối kỳ

**Các thành viên:**
1. MSSV: 18120143 - Họ tên: Nguyễn Đình Hoàng Phúc
2. MSSV: 18120512 - Họ tên: Lê Đặng Thiên Phúc

---

## Import thư viện

In [74]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

## Thu thập dữ liệu

Trong đồ án lần này nhóm chúng em sử dụng bộ dữ liệu thông tin các **Ứng dụng trên Google Play Store**. <br>
Bộ dữ liệu được tạo vào 05/04/19 và cập nhật lần cuối vào 16/06/21 trên trang Kaggle, [đường dẫn](https://www.kaggle.com/gauthamp10/google-playstore-apps). Bộ dữ liệu được thu thập bằng cách crawl từ trang web của Google Play sử dụng tool Scrapy chạy trên cloud VM và có chứa nhiều thông tin liên quan đến ứng dụng như tên, kích thước, rating,...

## Khám phá dữ liệu

### Đọc dữ liệu từ file .csv vào Dataframe

In [75]:
gps_df = pd.read_csv('Google-Playstore.csv')
gps_df.head()

App Name  \
0                                            Gakondo   
1                                Ampere Battery Info   
2                                             Vibook   
3  Smart City Trichy Public Service Vehicles 17UC...   
4                                            GROW.me   

                       App Id       Category  Rating  Rating Count Installs  \
0         com.ishakwe.gakondo      Adventure     0.0           0.0      10+   
1  com.webserveis.batteryinfo          Tools     4.4          64.0   5,000+   
2        com.doantiepvien.crm   Productivity     0.0           0.0      50+   
3     cst.stJoseph.ug17ucs548  Communication     5.0           5.0      10+   
4        com.horodyski.grower          Tools     0.0           0.0     100+   

   Minimum Installs  Maximum Installs  Free  Price  ...  \
0              10.0                15  True    0.0  ...   
1            5000.0              7662  True    0.0  ...   
2              50.0                58  True    0.0  ...   
3              10.0                19  True    0.0  ...   
4             100.0               478  True    0.0  ...   

                  Developer Website              Developer Email  \
0       https://beniyizibyose.tk/#/       jean21101999@gmail.com   
1   https://webserveis.netlify.app/         webserveis@gmail.com   
2                               NaN          vnacrewit@gmail.com   
3  http://www.climatesmarttech.com/  climatesmarttech2@gmail.com   
4       http://www.horodyski.com.pl    rmilekhorodyski@gmail.com   

       Released  Last Updated Content Rating  \
0  Feb 26, 2020  Feb 26, 2020       Everyone   
1  May 21, 2020  May 06, 2021       Everyone   
2   Aug 9, 2019  Aug 19, 2019       Everyone   
3  Sep 10, 2018  Oct 13, 2018       Everyone   
4  Feb 21, 2020  Nov 12, 2018       Everyone   

                                      Privacy Policy Ad Supported  \
0                 https://beniyizibyose.tk/projects/        False   
1  https://dev4phones.wordpress.com/licencia-de-uso/         True   
2  https://www.vietnamairlines.com/vn/en/terms-an...        False   
3                                                NaN         True   
4                        http://www.horodyski.com.pl        False   

  In App Purchases Editors Choice         Scraped Time  
0            False          False  2021-06-15 20:19:35  
1            False          False  2021-06-15 20:19:35  
2            False          False  2021-06-15 20:19:35  
3            False          False  2021-06-15 20:19:35  
4            False          False  2021-06-15 20:19:35  

[5 rows x 24 columns]

### Dữ liệu gồm bao nhiêu dòng và cột?

In [76]:
num_rows, num_cols = gps_df.shape
print('Bộ dữ liệu có {} dòng và {} cột'.format(num_rows, num_cols))

Bộ dữ liệu có 2312944 dòng và 24 cột


### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Mỗi dòng là chi tiết thông tin của một ứng dụng trên Google Play Store. Có vẻ không có dòng nào có ý nghĩa khác.

### Dữ liệu có các dòng bị lặp không?

In [77]:
duplicated_rows = gps_df.duplicated()
num_duplicated_rows = len(gps_df[duplicated_rows])
print("Có {} dòng bị lặp.".format(num_duplicated_rows))

Có 0 dòng bị lặp.


### Mỗi cột có ý nghĩa gì?

Dưới đây là phần mô tả của tác giả về các cột trong file "Google-Playstore.csv":

- **App Name**: tên ứng dụng.
- **App Id**:  id của ứng dụng.
- **Category**: thể loại ứng dụng.
- **Rating**: chỉ số đánh giá trung bình.
- **Rating Count**: số lượt đánh giá.
- **Installs**: số lượt tải.
- **Minimum Installs**: số lượt tải tối thiểu.
- **Maximum Installs**: số lượt tải tối đa.
- **Free**: ứng dụng là mĩễn phí hay trả tiền.
- **Price**: giá ứng dụng.
- **Currency**: đồng tiền dùng để thanh toán.
- **Size**: kích cỡ ứng dụng
- **Minimum Android**: phiên bản android tối thiểu để chạy ứng dụng.
- **Developer Id**: ID của nhà phát triển.
- **Developer Website**: Web của nhà phát triển.
- **Developer Email**: Email của nhà phát triển.
- **Released**: Ngày phát hành.
- **Last Updated**: Ngày cập nhật gần đây nhất.
- **Content Rating**: Đánh giá nội dung ứng dụng theo chính sách Google Play.
- **Privacy Policy**: Chính sách bảo mật.
- **Ad Supported**: Ứng dụng có quảng cáo không.
- **In app purchases**: Có thực hiện mua bán trong ứng dụng không.
- **Editor Choice**: Có được đánh giá chuẩn Editor Choice the Google Play không?
- **Scraped Time**: thời gian thu nhập dữ liệu (GMT)

### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

In [78]:
col_dtypes = gps_df.dtypes
col_dtypes

App Name              object
App Id                object
Category              object
Rating               float64
Rating Count         float64
Installs              object
Minimum Installs     float64
Maximum Installs       int64
Free                    bool
Price                float64
Currency              object
Size                  object
Minimum Android       object
Developer Id          object
Developer Website     object
Developer Email       object
Released              object
Last Updated          object
Content Rating        object
Privacy Policy        object
Ad Supported            bool
In App Purchases        bool
Editors Choice          bool
Scraped Time          object
dtype: object

- Các cột `Rating`, `Rating Count`, `Minium Installs`, `Maximum Installs` `Price` có dạng numerical.
- Các cột `Free`, `Ad Supported`, `In App Purchases`, `Editors Choice` có kiểu `bool` thuộc dạng categorical.
- Các cột còn lại đều có kiểu là `object`. Do không có cảnh báo về mixed typed. Dự đoán các cột này có kiểu dữ liệu thật là `string`. Ta quan sát kỹ thêm một chút các cột này để xử lý.

#### Quan sát thêm các cột có kiểu `object`

Các hàm hỗ trợ:

In [79]:
def missing_ratio(s):
    return s.isna().mean() * 100
def num_diff_vals(s):
    return s.nunique()
def diff_vals(s):
    return s.dropna().unique()

Quan sát các cột có kiểu dữ liệu `object`.

In [80]:
object_col_df = gps_df[['App Name', 'App Id', 'Category', 'Installs', 'Currency',
       'Size', 'Minimum Android', 'Developer Id', 'Developer Website',
       'Developer Email', 'Released', 'Last Updated', 'Content Rating',
       'Privacy Policy',
       'Scraped Time']]
object_col_profiles_df = object_col_df.agg([missing_ratio, num_diff_vals, diff_vals])
object_col_profiles_df

App Name  \
missing_ratio                                           0.000086   
num_diff_vals                                            2177944   
diff_vals      [Gakondo, Ampere Battery Info, Vibook, Smart C...   

                                                          App Id  \
missing_ratio                                                0.0   
num_diff_vals                                            2312944   
diff_vals      [com.ishakwe.gakondo, com.webserveis.batteryin...   

                                                        Category  \
missing_ratio                                                0.0   
num_diff_vals                                                 48   
diff_vals      [Adventure, Tools, Productivity, Communication...   

                                                        Installs  \
missing_ratio                                           0.004626   
num_diff_vals                                                 22   
diff_vals      [10+, 5,000+, 50+, 100+, 1,000+, 500+, 50,000+...   

                                                        Currency  \
missing_ratio                                           0.005837   
num_diff_vals                                                 15   
diff_vals      [USD, XXX, CAD, EUR, INR, VND, GBP, BRL, KRW, ...   

                                                            Size  \
missing_ratio                                           0.008474   
num_diff_vals                                               1657   
diff_vals      [10M, 2.9M, 3.7M, 1.8M, 6.2M, 46M, 2.5M, 16M, ...   

                                                 Minimum Android  \
missing_ratio                                           0.282324   
num_diff_vals                                                154   
diff_vals      [7.1 and up, 5.0 and up, 4.0.3 and up, 4.1 and...   

                                                    Developer Id  \
missing_ratio                                           0.001427   
num_diff_vals                                             758371   
diff_vals      [Jean Confident Irénée NIYIZIBYOSE, Webserveis...   

                                               Developer Website  \
missing_ratio                                          32.894657   
num_diff_vals                                             810440   
diff_vals      [https://beniyizibyose.tk/#/, https://webserve...   

                                                 Developer Email  \
missing_ratio                                            0.00134   
num_diff_vals                                             950456   
diff_vals      [jean21101999@gmail.com, webserveis@gmail.com,...   

                                                        Released  \
missing_ratio                                           3.071972   
num_diff_vals                                               4158   
diff_vals      [Feb 26, 2020, May 21, 2020, Aug 9, 2019, Sep ...   

                                                    Last Updated  \
missing_ratio                                                0.0   
num_diff_vals                                               3918   
diff_vals      [Feb 26, 2020, May 06, 2021, Aug 19, 2019, Oct...   

                                                  Content Rating  \
missing_ratio                                                0.0   
num_diff_vals                                                  6   
diff_vals      [Everyone, Teen, Mature 17+, Everyone 10+, Adu...   

                                                  Privacy Policy  \
missing_ratio                                          18.199879   
num_diff_vals                                             977743   
diff_vals      [https://beniyizibyose.tk/projects/, https://d...   

                                                    Scraped Time  
missing_ratio                                                0.0  
num_diff_vals                                              67374  
diff_vals      [2021-06-15 20:19:

Quan sát kỹ hơn cột `Minimum Android`

In [81]:
object_col_profiles_df['Minimum Android']['diff_vals']

array(['7.1 and up', '5.0 and up', '4.0.3 and up', '4.1 and up',
       '6.0 and up', '4.4 and up', '4.0 and up', '4.2 and up',
       '2.1 and up', '7.0 and up', '2.2 and up', '2.3 and up',
       '4.3 and up', '5.1 and up', '1.6 and up', '3.0 and up',
       'Varies with device', '8.0 and up', '2.3.3 and up', '4.4W and up',
       '3.2 and up', '1.5 and up', '4.0.3 - 7.1.1', '3.1 and up',
       '2.0 and up', '1.0 and up', '1.1 and up', '2.0.1 and up',
       '4.4 - 6.0', '2.1 - 4.4', '4.0 - 5.0', '2.0 - 2.3.4', '2.2 - 4.4',
       '4.1 - 7.0', '4.1 - 6.0', '4.0 - 4.4W', '4.0 - 4.4', '4.0.3 - 7.0',
       '2.3 - 5.0', '4.0 - 7.1.1', '4.1 - 7.1.1', '4.0 - 7.0',
       '4.1 - 8.0', '4.0 - 6.0', '2.3 - 7.0', '4.0 - 8.0', '2.3 - 6.0',
       '2.1 - 2.3.4', '3.0 - 4.1.1', '2.2 - 3.0', '2.0 - 8.0',
       '4.0.3 - 8.0', '3.0 - 4.4W', '4.4 - 7.1.1', '4.4 - 8.0',
       '4.0.3 - 6.0', '1.6 - 4.0.4', '8.0', '2.3 - 5.1', '2.1 - 7.1.1',
       '2.3 - 4.4W', '4.4 - 7.0', '6.0 - 7.1.1', '2.3 - 4.

Quan sát dữ liệu và dựa và ý nghĩa, ta thấy:
- Các cột `App Name`, `App Id`, `Category`, `Currency`, `Developer Id`, `Developer Website`, `Developer Email`,`Content Rating`, `Privacy Policy`, chắc chắn được xếp vào categorical. 
- Các cột `Installs`: số lượt tải, `Size`: kích thước ứng dụng có thể được đưa sang kiểu số.
- Các cột `Released`, `Last Updated`, `Scraped Time`: có thể chuyển về datetime.
- Cột `Minimum Android` : là dữ liệu dạng khoảng, có thể tiền xử lý chia thành hai cột phiên bản hệ điều hành thấp nhất và cao nhất. Hai cột mới thêm vào có thể chuyển về dạng numerical, nhưng ta sẽ chưa chuyển vội. Sẽ tiền xử lý tiếp nếu thấy cần thiết.

#### Tiền xử lý

Ta tiền xử lý chuyển dữ liệu các cột được phân loại là numerical có kiểu object về lại kiểu số.

##### Chuyển cột `Size` sang dạng số:
Xem thử các pattern đứng sau dữ liệu số:

In [82]:
set(gps_df['Size'].str.extract(r'([^\d]+)$', expand=False).values)

{'G', 'M', 'Varies with device', 'k', nan}

Ta chỉ chấp nhận các pattern string của cột `Size` có dạng là:
- Chứa chuỗi: *Varies with device*
- Kích thước dưới dạng: `{số}M` (số Megabyte của ứng dụng).
- Kích thước dưới dạng: `{số}k` (số Kilobyte của ứng dụng). 
- Kích thước dưới dạng: `{số}G` (số Gigabyte của ứng dụng). 
Đếm xem thử có bao nhiêu dòng không có dạng như vậy:

In [83]:
out_of_pattern_rows = gps_df['Size'].str.replace(',','').str.extract(r'((?:Varies with device)|(?:\d+.\d+[MGk])|(?:\d+[MGk]))', expand=False).isna()
print("Số dòng không chuyển đổi được: ", len(gps_df[out_of_pattern_rows]))
print("Các dòng không chuyển đổi được có các giá trị: ",gps_df[out_of_pattern_rows]['Size'].unique())

Số dòng không chuyển đổi được:  196
Các dòng không chuyển đổi được có các giá trị:  [nan]


Chỉ có **196** dòng lạc loài do chúng mang giá trị "nan" cũng không đáng kể.

Mặc dù ý nghĩa của 'Varies with device' và không theo pattern là khác nhau. Trong đồ án này nhóm chúng em xem hai giá trị này là tương đương nhau là không thu thập được và để giá trị `N/A` cho chúng.

In [84]:
gps_df.loc[out_of_pattern_rows,'Size'] = np.nan
gps_df.loc[gps_df['Size']=='Varies with device','Size'] = np.nan

Tách phần số của cột `Size`

In [85]:
size_nums = gps_df['Size'].str.replace(',','').str.extract(r'((?:\d+.\d+)|(?:\d+))', expand=False).astype('float64')

Các dòng có đơn vị là KB và GB cần được chuyển đổi về cùng đơn vị MB.

In [86]:
size_nums[gps_df['Size'].str[-1]=='k'] *= (1/1024)
size_nums[gps_df['Size'].str[-1]=='G'] *= 1024

Gán lại vào Dataframe chính.

In [87]:
gps_df['Size'] = size_nums

##### Chuyển cột `Installs` sang dạng số: 
Ta chỉ chấp nhận các pattern string của cột `Installs` có dạng là: số nguyên và dấu cộng "+" liền sau đó. (1000+, 1000,000+,...). Đếm xem thử có bao nhiêu dòng không có dạng như vậy:

In [88]:
out_of_pattern_rows = gps_df['Installs'].str.replace(',','').str.extract(r'((?:\d+[+]))', expand=False).isna()
print("Số dòng không chuyển đổi được: ", len(gps_df[out_of_pattern_rows]))
print("Các dòng không chuyển đổi được có các giá trị: ",gps_df[out_of_pattern_rows]['Installs'].unique())

Số dòng không chuyển đổi được:  107
Các dòng không chuyển đổi được có các giá trị:  [nan]


Chỉ có **107** dòng lạc loài, cũng không đáng kể, ta bỏ qua các dòng này và cho chúng bằng `N/A`.

In [89]:
gps_df.loc[out_of_pattern_rows,'Installs'] = np.nan

Gán lại vào Dataframe chính:

In [90]:
install_nums = gps_df['Installs'].str.replace(',','').str.extract(r'((?:\d+))', expand=False).astype('float64')
gps_df['Installs'] = install_nums

##### Chuyển các cột `Released`, `Last Updated`, `Scraped Time` sang datetime.

Làm với cột `Released` trước:

In [91]:
out_of_pattern_rows = pd.to_datetime(gps_df['Released'], format = '%b %d, %Y', errors = 'coerce').isna()
print("Số dòng không chuyển đổi được: ", len(gps_df[out_of_pattern_rows]))
print("Các dòng không chuyển đổi được có các giá trị: ",gps_df[out_of_pattern_rows]['Released'].unique())

Số dòng không chuyển đổi được:  71053
Các dòng không chuyển đổi được có các giá trị:  [nan]


Có **71035** dòng dữ liệu không chuyển được. Cũng khá nhiều nhưng đều mang giá trị `N/A` nên có thể bỏ qua

In [92]:
gps_df['Released'] = pd.to_datetime(gps_df['Released'], format = '%b %d, %Y', errors = 'coerce')

Làm với cột `Last Updated`:

In [93]:
out_of_pattern_rows = pd.to_datetime(gps_df['Last Updated'], format = '%b %d, %Y', errors = 'coerce').isna()
print("Số dòng không chuyển đổi được: ", len(gps_df[out_of_pattern_rows]))
print("Các dòng không chuyển đổi được có các giá trị: ",gps_df[out_of_pattern_rows]['Last Updated'].unique())

Số dòng không chuyển đổi được:  0
Các dòng không chuyển đổi được có các giá trị:  []


In [94]:
gps_df['Last Updated'] = pd.to_datetime(gps_df['Last Updated'], format = '%b %d, %Y', errors = 'coerce')

Làm với cột `Scraped Time`:

In [95]:
out_of_pattern_rows = pd.to_datetime(gps_df['Scraped Time'], format = '%Y-%m-%d %H:%M:%S', errors = 'coerce').isna()
print("Số dòng không chuyển đổi được: ", len(gps_df[out_of_pattern_rows]))
print("Các dòng không chuyển đổi được có các giá trị: ",gps_df[out_of_pattern_rows]['Scraped Time'].unique())

Số dòng không chuyển đổi được:  0
Các dòng không chuyển đổi được có các giá trị:  []


In [96]:
gps_df['Scraped Time'] = pd.to_datetime(gps_df['Scraped Time'], format = '%Y-%m-%d %H:%M:%S', errors = 'coerce')

Tới đây ta đã giải quyết xong một số cột có dữ liệu `numerical` cần thiết. <br>
##### Kiểm tra xem dữ liệu `numerical` đã hợp lý chưa?

In [97]:
numerical_col_df = gps_df[['Rating', 'Rating Count', 'Installs', 'Size' ,'Minimum Installs', 'Maximum Installs', 'Price', 'Released', 'Last Updated', 'Scraped Time']]
numerical_col_profiles_df = numerical_col_df.agg([missing_ratio, pd.Series.min, pd.Series.max])
numerical_col_profiles_df

Rating  Rating Count      Installs         Size  \
missing_ratio  0.989345  9.893452e-01  4.626139e-03     3.241453   
min            0.000000  0.000000e+00  0.000000e+00     0.003125   
max            5.000000  1.385576e+08  1.000000e+10  1536.000000   

               Minimum Installs  Maximum Installs  Price             Released  \
missing_ratio      4.626139e-03      0.000000e+00    0.0             3.071972   
min                0.000000e+00      0.000000e+00    0.0  2010-01-28 00:00:00   
max                1.000000e+10      1.205763e+10  400.0  2021-06-16 00:00:00   

                      Last Updated         Scraped Time  
missing_ratio                  0.0                  0.0  
min            2009-02-09 00:00:00  2021-06-15 20:19:35  
max            2021-06-16 00:00:00  2021-06-16 15:10:42

Có vẻ chưa có gì bất thường

Ta sẽ tiếp tục xử lý cho các cột khác!

##### Tách cột `Minimum Android` thành 2 cột là Min Android Ver và Max Android Ver

In [98]:
android_ver_pattern_1 = gps_df['Minimum Android'].str.extract(r'((?:\d.\d.\d)|(?:\d.\d))', expand = False)
android_ver_pattern_2 = gps_df['Minimum Android'].str.extract(r'((?<=\- ).*)', expand=False)

In [99]:
gps_df['Min Android Ver'] = android_ver_pattern_1

In [100]:
gps_df['Max Android Ver'] = android_ver_pattern_2

Tới đây ta đã giải quyết xong một số cột có dữ liệu `categorical` cần thiết. 
##### Kiểm tra xem dữ liệu `categorical` đã hợp lý chưa?

In [101]:
categorical_col_df = gps_df[['App Name', 'App Id', 'Category', 'Currency',
                             'Minimum Android', 'Min Android Ver', 'Max Android Ver', 'Developer Id', 'Developer Website',
                             'Developer Email', 'Content Rating',
                             'Privacy Policy', 'Free', 'Ad Supported', 'In App Purchases', 'Editors Choice']]
categorical_col_profiles_df = categorical_col_df.agg([missing_ratio, num_diff_vals, diff_vals])
categorical_col_profiles_df

App Name  \
missing_ratio                                           0.000086   
num_diff_vals                                            2177944   
diff_vals      [Gakondo, Ampere Battery Info, Vibook, Smart C...   

                                                          App Id  \
missing_ratio                                                0.0   
num_diff_vals                                            2312944   
diff_vals      [com.ishakwe.gakondo, com.webserveis.batteryin...   

                                                        Category  \
missing_ratio                                                0.0   
num_diff_vals                                                 48   
diff_vals      [Adventure, Tools, Productivity, Communication...   

                                                        Currency  \
missing_ratio                                           0.005837   
num_diff_vals                                                 15   
diff_vals      [USD, XXX, CAD, EUR, INR, VND, GBP, BRL, KRW, ...   

                                                 Minimum Android  \
missing_ratio                                           0.282324   
num_diff_vals                                                154   
diff_vals      [7.1 and up, 5.0 and up, 4.0.3 and up, 4.1 and...   

                                                 Min Android Ver  \
missing_ratio                                           2.280384   
num_diff_vals                                                 25   
diff_vals      [7.1, 5.0, 4.0.3, 4.1, 6.0, 4.4, 4.0, 4.2, 2.1...   

                                                 Max Android Ver  \
missing_ratio                                          99.980415   
num_diff_vals                                                 18   
diff_vals      [7.1.1, 6.0, 4.4, 5.0, 2.3.4, 7.0, 4.4W, 8.0, ...   

                                                    Developer Id  \
missing_ratio                                           0.001427   
num_diff_vals                                             758371   
diff_vals      [Jean Confident Irénée NIYIZIBYOSE, Webserveis...   

                                               Developer Website  \
missing_ratio                                          32.894657   
num_diff_vals                                             810440   
diff_vals      [https://beniyizibyose.tk/#/, https://webserve...   

                                                 Developer Email  \
missing_ratio                                            0.00134   
num_diff_vals                                             950456   
diff_vals      [jean21101999@gmail.com, webserveis@gmail.com,...   

                                                  Content Rating  \
missing_ratio                                                0.0   
num_diff_vals                                                  6   
diff_vals      [Everyone, Teen, Mature 17+, Everyone 10+, Adu...   

                                                  Privacy Policy  \
missing_ratio                                          18.199879   
num_diff_vals                                             977743   
diff_vals      [https://beniyizibyose.tk/projects/, https://d...   

                        Free   Ad Supported In App Purchases Editors Choice  
missing_ratio            0.0            0.0              0.0            0.0  
num_diff_vals              2              2                2              2  
diff_vals      [True, False]  [False, True]    [False, True]  [False, True]

Quan sát kỹ hơn 4 cột `Category`, `Currency` và `Content Rating`

In [102]:
categorical_col_profiles_df['Category']['diff_vals']

array(['Adventure', 'Tools', 'Productivity', 'Communication', 'Social',
       'Libraries & Demo', 'Lifestyle', 'Personalization', 'Racing',
       'Maps & Navigation', 'Travel & Local', 'Food & Drink',
       'Books & Reference', 'Medical', 'Puzzle', 'Entertainment',
       'Arcade', 'Auto & Vehicles', 'Photography', 'Health & Fitness',
       'Education', 'Shopping', 'Board', 'Music & Audio', 'Sports',
       'Beauty', 'Business', 'Educational', 'Finance', 'News & Magazines',
       'Casual', 'Art & Design', 'House & Home', 'Card', 'Events',
       'Trivia', 'Weather', 'Strategy', 'Word', 'Video Players & Editors',
       'Action', 'Simulation', 'Music', 'Dating', 'Role Playing',
       'Casino', 'Comics', 'Parenting'], dtype=object)

In [103]:
categorical_col_profiles_df['Currency']['diff_vals']

array(['USD', 'XXX', 'CAD', 'EUR', 'INR', 'VND', 'GBP', 'BRL', 'KRW',
       'TRY', 'RUB', 'SGD', 'AUD', 'PKR', 'ZAR'], dtype=object)

In [104]:
categorical_col_profiles_df['Content Rating']['diff_vals']

array(['Everyone', 'Teen', 'Mature 17+', 'Everyone 10+',
       'Adults only 18+', 'Unrated'], dtype=object)

Cột currency có giá trị "XXX" mang ý nghĩa là "No currency" (theo Wikipedia). Còn lại chưa có gì đáng chú ý hay bất thường.

## Đưa ra các câu hỏi có ý nghĩa cần trả lời

1. Có phải càng gần đây app càng nặng.
2. Thường các app được đưa ra vào tháng nào trong năm.
3. Các công ty có nhiều lượt tải nhất thì app chủ yếu chủ đề gì?
4. App có category nào sẽ có lượt tải xuống cao nhất (xét trên columns Install)?
5. Developer nào có tổng số lượt tải các app cao nhất?
6. Tương quan rating lượt tải.
7. Xu hướng app theo free hay tính phí?

## Phân tích dữ liệu để trả lời các câu hỏi trên

## Nhìn lại quá trình làm đồ án

### Thành viên: Nguyễn Đình Hoàng Phúc

### Thành viên: Lê Đặng Thiên Phúc

## Tài liệu tham khảo